# ***Functions Assignment***

---


1. What is the difference between a function and a method in Python?

Ans ->

Function: A block of reusable code that is not tied to an object. Defined using def.

Method: A function that is associated with an object (typically a class instance).

✅ Example:

In [5]:
# Function
def greet():
    return "Hello"

# Method
my_list = [1, 2, 3]
my_list.append(4)  # append is a method


2. Explain the concept of function arguments and parameters in Python.

Ans ->

Parameter: A variable in a function definition.

Argument: A value passed to the function when it is called.

✅ Example:

In [2]:
def add(a, b):  # a and b are parameters
    return a + b

add(2, 3)       # 2 and 3 are arguments


5

3. What are the different ways to define and call a function in Python?

Ans ->

Ways to define:
Using def keyword.

Using lambda expression.

Ways to call:
Direct call by name.

Via function pointer.

Using *args and **kwargs.

✅ Example:




In [3]:
def greet(name):        # defining
    return "Hi " + name

greet("Dhiraj")         # direct call

f = greet
f("Ravi")               # function pointer


'Hi Ravi'

4. What is the purpose of the "return statement in a Python function?

It sends a value back to the caller and terminates the function's execution.

✅ Example:

In [6]:
def square(x):
    return x * x

result = square(4)  # returns 16


5. What are iterators in Python and how do they differ from iterables?

Ans ->

Iterable: Can be looped over (e.g., list, tuple).

Iterator: Object that keeps state and knows how to fetch the next item.

✅ Example:

In [9]:
numbers = [1, 2, 3]  # iterable
it = iter(numbers)   # now an iterator

print(next(it))      # Output: 1


1


6. Explain the concept of generators in Python and how they are defined.

Ans ->

Generators are functions that yield values one at a time.

Use yield instead of return.

✅ Example:

In [10]:
def count_up_to(n):
    i = 1
    while i <= n:
        yield i
        i += 1

for val in count_up_to(3):
    print(val)


1
2
3


7. What are the advantages of using generators over regular functions?

Ans ->

Memory-efficient: Yield items one at a time.

Lazy evaluation: Only compute when needed.

Suitable for large datasets.

✅ Example:

In [11]:
def generate_numbers():
    for i in range(1000000):
        yield i


8. What is a lambda function in Python and when is it typically used?

Ans ->

A small anonymous function defined using lambda.

Used for short-term use, especially with map(), filter(), etc.

✅ Example:

In [12]:
square = lambda x: x * x
print(square(5))  # Output: 25


25


9. Explain the purpose and usage of the 'map() function in Python.

Ans ->

Applies a function to each item in an iterable.

Returns a map object (needs to be converted to a list or iterated).

✅ Example:

In [15]:
nums = [1, 2, 3]
squares = list(map(lambda x: x**2, nums))  # [1, 4, 9]


10. What is the difference between map()', 'reduce()", and "filter() functions in Python?

Ans ->

| Function   | Purpose                                            | Returns      |
| ---------- | -------------------------------------------------- | ------------ |
| `map()`    | Applies a function to all items in an iterable     | New iterable |
| `reduce()` | Applies a rolling computation (e.g., sum, product) | Single value |
| `filter()` | Filters items that meet a condition                | New iterable |


In [17]:
from functools import reduce

nums = [1, 2, 3, 4]

# map
squares = list(map(lambda x: x**2, nums))  # [1, 4, 9, 16]

# reduce
total = reduce(lambda x, y: x + y, nums)   # 10

# filter
evens = list(filter(lambda x: x % 2 == 0, nums))  # [2, 4]


11. Using pen & Paper write the internal mechanism for sum operation using reduce function on this given list: [47,11,42,13]:

In [18]:
#You are performing:

from functools import reduce

reduce(lambda x, y: x + y, [47, 11, 42, 13])

113

Step-by-step:
First call: x = 47, y = 11 → result = 58

Second call: x = 58, y = 42 → result = 100

Third call: x = 100, y = 13 → result = 113

✅ Final result: 113

---



# ***Practical Questions:***

---



1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

In [19]:
def sum_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example
print(sum_even_numbers([1, 2, 3, 4, 5, 6]))  # Output: 12


12


2. Create a Python function that accepts a string and returns the reverse of that string.

In [20]:
def reverse_string(s):
    return s[::-1]

# Example
print(reverse_string("Dhiraj"))  # Output: jarihD


jarihD


3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [21]:
def square_list(numbers):
    return [x**2 for x in numbers]

# Example
print(square_list([1, 2, 3]))  # Output: [1, 4, 9]


[1, 4, 9]


4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [22]:
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False
    return True

# Print all primes from 1 to 200
primes = [x for x in range(1, 201) if is_prime(x)]
print(primes)


[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199]


5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [23]:
class FibonacciIterator:
    def __init__(self, max_terms):
        self.max_terms = max_terms
        self.a, self.b = 0, 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.max_terms:
            raise StopIteration
        self.count += 1
        self.a, self.b = self.b, self.a + self.b
        return self.a

# Example
for num in FibonacciIterator(7):
    print(num)


1
1
2
3
5
8
13


6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [24]:
def powers_of_two(max_exponent):
    for i in range(max_exponent + 1):
        yield 2 ** i

# Example
for power in powers_of_two(5):
    print(power)


1
2
4
8
16
32


7. implement a generator function that reads a file line by line and yields each line as a string.

In [25]:
def read_lines(filename):
    with open(filename, 'r') as file:
        for line in file:
            yield line.strip()

# Example
# for line in read_lines('example.txt'):
#     print(line)


8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [26]:
data = [(1, 3), (2, 1), (4, 2)]
sorted_data = sorted(data, key=lambda x: x[1])
print(sorted_data)  # Output: [(2, 1), (4, 2), (1, 3)]


[(2, 1), (4, 2), (1, 3)]


9. Write a Python program that uses 'map() to convert a list of temperatures from Celsius to Fahrenheit.

In [27]:
celsius = [0, 20, 37, 100]
fahrenheit = list(map(lambda c: (c * 9/5) + 32, celsius))
print(fahrenheit)  # Output: [32.0, 68.0, 98.6, 212.0]


[32.0, 68.0, 98.6, 212.0]


10. Create a Python program that uses filter() to remove all the vowels from a given string.

In [28]:
def remove_vowels(s):
    return ''.join(filter(lambda x: x.lower() not in 'aeiou', s))

print(remove_vowels("Dhiraj Shingane"))  # Output: Dhrj Shngn


Dhrj Shngn


11. Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:



Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the product of the price per item and the quantity. The product should be increased by 10.-C if the value of the order is smaller than 100,00 €.

In [29]:
orders = [
    (34587, "Learning Python, Mark Lutz", 4, 40.95),
    (98762, "Programming Python, Mark Lutz", 5, 56.80),
    (77226, "Head First Python, Paul Barry", 3, 32.95),
    (88112, "Einführung in Python3, Bernd Klein", 3, 24.99)
]

result = list(map(
    lambda order: (
        order[0],
        round(order[2] * order[3] + (10 if order[2] * order[3] < 100 else 0), 2)
    ),
    orders
))

print(result)


[(34587, 163.8), (98762, 284.0), (77226, 108.85), (88112, 84.97)]
